### PR TITLE
Add RPC Builder to Substrate Node Template

### PR DESCRIPTION
**Pull request by Meuko**
Allows for and gives a concrete example on how to implement custom RPCs (with its implementation in the template's service.rs file). I've tried to keep everything minimal, but since I don't understand everything I can almost guarantee that there's room for improvements- not sure if allowed but I added 2 TODO's stating where these can be added. Note that the builder is only implemented for the full client. I've fully tested this and it works, perhaps if needed- could somebody point me towards some documentation which shows me how to write test-cases for RPCs? Thanks and sorry if I made a mess.

### CODE REVIEW

File: bin/node-template/node/Cargo.toml [hunk 1]
```diff
@@ -31,9 +32,16 @@ sp-consensus = { version = "0.8.0-rc5", path = "../../../primitives/consensus/co
...
+sp-api = { version = "2.0.0-rc5", path = "../../../primitives/api" }
+sc-rpc-api = { version = "0.8.0-rc5", path = "../../../client/rpc-api" }
 sc-client-api = { version = "2.0.0-rc5", path = "../../../client/api" }
+
 sp-runtime = { version = "2.0.0-rc5", path = "../../../primitives/runtime" }
-sc-basic-authorship = { path = "../../../client/basic-authorship", version = "0.8.0-rc5"}
+sp-blockchain = { version = "2.0.0-rc5", path = "../../../primitives/blockchain" }
+sp-block-builder = { version = "2.0.0-rc5", path = "../../../primitives/block-builder" }
+sc-basic-authorship = { version = "0.8.0-rc5", path = "../../../client/basic-authorship" }
+substrate-frame-rpc-system = { version = "2.0.0-rc5", path = "../../../utils/frame/rpc/system" }
```
Comment (shawntabrizi) [inline bin/node-template/node/Cargo.toml:44]: I recommend you group all the new dependencies into a seperate block below the existing dependencies, and label that block with a comment `# imports for rpc` or something like that

File: bin/node-template/node/src/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:10]: License in node-template should be unlicense. Copy from another file

File: bin/node-template/node/src/rpc.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+use node_template_runtime::{opaque::Block, AccountId, Index, Hash};
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::{Error as BlockChainError, HeaderMetadata, HeaderBackend};
+use sp_consensus::SelectChain;
+use sp_block_builder::BlockBuilder;
+pub use sc_rpc_api::DenyUnsafe;
+use sp_transaction_pool::TransactionPool;
+
+/// Light client extra dependencies.
+#[allow(dead_code)]
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:44]: Did you add this allow?

File: bin/node-template/node/src/rpc.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		pool,
+		select_chain: _,
+		deny_unsafe,
+	} = deps;
+
+	io.extend_with(
+		SystemApi::to_delegate(FullSystem::new(client.clone(), pool, deny_unsafe))
+	);
+
+        //io.extend_with(NodeTemplateApi::to_delegate(....
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:94]: Probably best to have a bit more comment here to be friendly to users

cc @danforbes

File: bin/node-template/node/src/rpc.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        io
+}
+
+/// Instantiate all Light RPC extensions.
+#[allow(dead_code)]
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:100]: Does node-template use light rpc?

if not, probably better to remove it

File: bin/node-template/node/src/service.rs [hunk 6]
```diff
@@ -18,6 +18,11 @@ native_executor_instance!(
 	node_template_runtime::native_version,
 );
 
+
+// TODO:: Please change this to crate, lrpc=localrpc
```
Comment (shawntabrizi) [inline bin/node-template/node/src/service.rs:22]: TODO?

File: bin/node-template/node/src/service.rs [hunk 7]
```diff
@@ -109,14 +114,32 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
 	let prometheus_registry = config.prometheus_registry().cloned();
 	let telemetry_connection_sinks = sc_service::TelemetryConnectionSinks::default();
 
+
```
Comment (shawntabrizi) [inline bin/node-template/node/src/service.rs:117]: extra newline?

File: bin/node-template/node/src/service.rs [hunk 8]
```diff
@@ -18,6 +18,11 @@ native_executor_instance!(
 	node_template_runtime::native_version,
 );
 
+
+// TODO:: Please change this to crate, lrpc=localrpc
+#[path = "rpc.rs"]
```
Comment (shawntabrizi) [inline bin/node-template/node/src/service.rs:23]: ??

**Comment by shawntabrizi** [comment]
@Meuko thanks for a great start. Take a look at some of my comments

File: bin/node-template/node/src/rpc.rs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+/// Full client dependencies.
+pub struct FullDeps<C, P, SC> {
+	/// The client instance to use.
+	pub client: Arc<C>,
+	/// Transaction pool instance.
+	pub pool: Arc<P>,
+	/// The SelectChain Strategy
+	pub select_chain: SC,
```
Comment (g2udevelopment) [inline bin/node-template/node/src/rpc.rs:63]: select_chain is not used an needed, consider removing

File: bin/node-template/runtime/Cargo.toml [hunk 10]
```diff
@@ -13,6 +13,8 @@ targets = ["x86_64-unknown-linux-gnu"]
 [dependencies]
 codec = { package = "parity-scale-codec", version = "1.3.4", default-features = false, features = ["derive"] }
 
+system-runtime-rpi = { default-features = false, package = "frame-system-rpc-runtime-api" , version = "2.0.0-rc5", path = "../../../frame/system/rpc/runtime-api/" }
```
Comment (g2udevelopment) [inline bin/node-template/runtime/Cargo.toml:16]: Is this a typo , or did you intentionly used rpi instead of api?

Comment (Meuko) [reply bin/node-template/runtime/Cargo.toml:16]: Intentional, my way of condensing the name- I'll probably change it back to be more consistent with the rest of the project, my apologies, thanks!

File: bin/node-template/node/src/rpc.rs (hunk 9 above)
Comment (Meuko) [reply bin/node-template/node/src/rpc.rs:63]: Indeed, I'll clean that up too. Thanks!

File: bin/node-template/node/src/service.rs (hunk 8 above)
Comment (Meuko) [reply bin/node-template/node/src/service.rs:23]: I wasn't sure whether or not it was ideal to have the whole rpc.rs loaded into service.rs. If I recall correctly the original, in the node- crate::... is used, which would error out for me. Perhaps more reading on namespaces in Rust for me.

File: bin/node-template/node/src/rpc.rs (hunk 3 above)
Comment (Meuko) [reply bin/node-template/node/src/rpc.rs:44]: I did, I'll remove it, including the light blocks. I understand that dead code throws errors during the building process, so I tried and found some other dead_code declarations in the repo, so I assumed it was allowed, my bad, thanks!

File: bin/node-template/node/src/service.rs (hunk 8 above)
Comment (shawntabrizi) [reply bin/node-template/node/src/service.rs:23]: Cant we treat `rpc.rs` as a traditional rust module? I am wondering why this attribute is needed

File: bin/node-template/runtime/Cargo.toml (hunk 10 above)
Comment (shawntabrizi) [reply bin/node-template/runtime/Cargo.toml:16]: yeah, lets avoid renaming things :)

File: bin/node-template/node/src/rpc.rs (hunk 4 above)
Comment (Meuko) [reply bin/node-template/node/src/rpc.rs]: Think I need some more help on how to give a concrete explanation on what's happening there. I did my best with my current explanation but feel free to tell me what to write there. Thanks.

File: bin/node-template/node/src/rpc.rs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+		client,
+		pool,
+		deny_unsafe,
+	} = deps;
+
+	io.extend_with(
+		SystemApi::to_delegate(FullSystem::new(client.clone(), pool, deny_unsafe))
+	);
+
+        // To expand upon the current RPC set using a custom written API, extend using the syntaxt
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:60]:
```suggestion
        // To expand upon the current RPC set using a custom written API, extend using the syntax
```

File: bin/node-template/node/src/rpc.rs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        // listed below. YourRpcStruct is the struct with a reference to a client, which is needed
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:61]:
```suggestion
        // listed below. `YourRpcStruct` is the struct with a reference to a client, which is needed
```

File: bin/node-template/node/src/rpc.rs [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        // to call into the runtime. More information on this can be found in the "Custom RPCs"
+        // recipe.
+        //io.extend_with(YourRpcTrait::to_delegate(YourRpcStruct::new(ReferenceToClient, ...)));
```
Comment (shawntabrizi) [inline bin/node-template/node/src/rpc.rs:64]:
```suggestion
        // To expand upon the current RPC set using a custom written API, extend using the syntax
        // listed below. `YourRpcStruct` is the struct with a reference to a client, which is needed
        // to call into the runtime.
        //io.extend_with(YourRpcTrait::to_delegate(YourRpcStruct::new(ReferenceToClient, ...)));
```

**Comment by Meuko** [comment]
> I think there are issues between spaces and tabs here. Please make sure everything uses tabs.
> 
> Other than that this looks great! Thank you @Meuko

Hey thanks @shawntabrizi , I'm not sure which files you're referring to, is it all my committed files?

File: bin/node-template/node/Cargo.toml [hunk 14]
```diff
@@ -33,7 +33,17 @@ sc-finality-grandpa = { version = "0.8.0-rc5", path = "../../../client/finality-
 sp-finality-grandpa = { version = "2.0.0-rc5", path = "../../../primitives/finality-grandpa" }
 sc-client-api = { version = "2.0.0-rc5", path = "../../../client/api" }
 sp-runtime = { version = "2.0.0-rc5", path = "../../../primitives/runtime" }
-sc-basic-authorship = { path = "../../../client/basic-authorship", version = "0.8.0-rc5"}
+
+
+# The dependencies listed below are used for the node-template's RPCs
```
Comment (danforbes) [inline bin/node-template/node/Cargo.toml:38]:
```suggestion

# These dependencies are used for the node template's RPCs
```

File: bin/node-template/node/src/rpc.rs [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+//! Since `substrate` core functionality makes no assumptions
+//! about the modules used inside the runtime, so do
+//! RPC methods defined in `sc-rpc` crate.
+//! It means that `client/rpc` can't have any methods that
+//! need some strong assumptions about the particular runtime.
+//!
+//! The RPCs available in this crate however can make some assumptions
+//! about how the runtime is constructed and what FRAME pallets
+//! are part of it. Therefore all node-runtime-specific RPCs can
+//! be placed here or imported from corresponding FRAME RPC definitions.
```
Comment (danforbes) [inline bin/node-template/node/src/rpc.rs:12]:
```suggestion
//! Substrate provides the `sc-rpc` crate, which defines the core RPC layer
//! used by Substrate nodes. This file extends those RPC definitions with
//! capabilities that are specific to this project's runtime configuration.
```

File: bin/node-template/node/src/rpc.rs [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+pub struct FullDeps<C, P> {
+	/// The client instance to use.
+	pub client: Arc<C>,
+	/// Transaction pool instance.
+	pub pool: Arc<P>,
+	/// Whether to deny unsafe calls
+	pub deny_unsafe: DenyUnsafe,
+}
+
+/// Instantiate all Full RPC extensions.
```
Comment (danforbes) [inline bin/node-template/node/src/rpc.rs:36]:
```suggestion
/// Instantiate all full RPC extensions.
```

File: bin/node-template/node/src/rpc.rs [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+	} = deps;
+
+	io.extend_with(
+		SystemApi::to_delegate(FullSystem::new(client.clone(), pool, deny_unsafe))
+	);
+
+	// To expand upon the current RPC set using a custom written API, extend using the syntax
+	// listed below. `YourRpcStruct` is the struct with a reference to a client, which is needed
+	// to call into the runtime.
+	//io.extend_with(YourRpcTrait::to_delegate(YourRpcStruct::new(ReferenceToClient, ...)));
```
Comment (danforbes) [inline bin/node-template/node/src/rpc.rs:63]:
```suggestion
	// Extend this RPC with a custom API by using the following syntax.
	// `YourRpcStruct` should have a reference to a client, which is needed
	// to call into the runtime.
	// `io.extend_with(YourRpcTrait::to_delegate(YourRpcStruct::new(ReferenceToClient, ...)));`
```

File: bin/node-template/runtime/Cargo.toml [hunk 18]
```diff
@@ -36,6 +35,9 @@ sp-std = { version = "2.0.0-rc5", default-features = false, path = "../../../pri
 sp-transaction-pool = { version = "2.0.0-rc5", default-features = false, path = "../../../primitives/transaction-pool" }
 sp-version = { version = "2.0.0-rc5", default-features = false, path = "../../../primitives/version" }
 
+# The dependency listed below is used for node-template's RPCs
```
Comment (danforbes) [inline bin/node-template/runtime/Cargo.toml:38]:
```suggestion
# Used for the node template's RPCs
```

**Comment by Meuko** [comment]
> Thank you so much, @Meuko, @g2udevelopment and @shawntabrizi 🙏

Thank you for the changes! Later tonight I'll push the final changes, cheers!

**Comment by Meuko** [comment]
Here we are, the updated ( docs have to reflect this ) transaction payment api is now implemented. @shawntabrizi sorry, but I might need some help with tabs and spaces again. I generally use RustFmt to get rid of them all, but my settings are radically different from what you guys seem to use, combine that with no rustfmt.toml. file and me hitting :RustFmt and the whole file will get a facelift XD. Thanks!

**Comment by thiolliere** [comment]
In vim you can show tab using `set listchars+=tab:>-`
and you can use tab automatically using `set noexpandtab`

File: bin/node-template/node/src/rpc.rs [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+		client,
+		pool,
+		deny_unsafe,
+	} = deps;
+
+	io.extend_with(
+		SystemApi::to_delegate(FullSystem::new(client.clone(), pool, deny_unsafe))
+	);
+
+        io.extend_with(
```
Comment (thiolliere) [inline bin/node-template/node/src/rpc.rs:54]: here it needs tab

File: bin/node-template/runtime/src/lib.rs [hunk 20]
```diff
@@ -417,4 +417,18 @@ impl_runtime_apis! {
...
+	impl frame_system_rpc_runtime_api::AccountNonceApi<Block, AccountId, Index> for Runtime {
+		fn account_nonce(account: AccountId) -> Index {
+			System::account_nonce(account)
+		}
+	}	
+           // crate::generic::block::traits::Extrinsic
+           // frame_support::inherent::Extrinsic
+           // sp_runtime::traits::Extrinsic
+        impl pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance> for Runtime {
+            fn query_info(uxt: <Block as BlockT>::Extrinsic , len: u32) -> pallet_transaction_payment_rpc_runtime_api::RuntimeDispatchInfo<Balance> {
```
Comment (thiolliere) [inline bin/node-template/runtime/src/lib.rs:430]: and here too

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
@Meuko 🙇

**Pull request merged**